clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -354,6 +358,460 @@ fatbinary(ArrayRef<std::pair<StringRef, StringRef>> InputFiles,
+      Files.push_back(Entry.IRFile);
+    }
+    return Files;
+  }
+
+  Expected<StringRef> writeSYCLTableToFile(void) {
+    // Create a new file.
+    auto TempFileOrErr =
+        createOutputFile(sys::path::filename(ExecutableName), "table");
+    if (!TempFileOrErr)
+      return TempFileOrErr.takeError();
+    std::error_code EC;
+    raw_fd_ostream TableFile(*TempFileOrErr, EC, sys::fs::OF_None);
+    if (EC)
+      reportError(errorCodeToError(EC));
+    TableFile << "[Code|Properties|Symbols]\n";
+    for (auto Entry : Entries) {
+      TableFile << Entry.IRFile << "|";
+      TableFile << Entry.PropFile << "|";
+      TableFile << Entry.SymFile << "\n";
+    }
+    return *TempFileOrErr;
+  }
+
+  Error populateSYCLTable(StringRef EntriesFile) {
+    llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> MBOrErr =
+        llvm::MemoryBuffer::getFileOrSTDIN(EntriesFile);
+    if (std::error_code EC = MBOrErr.getError())
+      return createFileError(EntriesFile, EC);
+    int LineNumber = -1;
+    for (line_iterator LI(**MBOrErr); !LI.is_at_eof(); ++LI) {
+      // Skip first line
+      StringRef Line = *LI;
+      if (LineNumber == -1) {
+        if (Line != "[Code|Properties|Symbols]")
+          return createStringError(inconvertibleErrorCode(),
+                                   "Invalid SYCL Table file.");
+        LineNumber++;
+        continue;
+      }
+      if (Line.empty())
+        return createStringError(inconvertibleErrorCode(),
+                                 "Invalid SYCL Table file.");
+      auto [FirstWord, Rem1] = Line.split("|");
+      SYCLTableEntry Entry;
+      Entry.IRFile = FirstWord.str();
+      if (Rem1.empty())
+        return createStringError(inconvertibleErrorCode(),
+                                 "Invalid SYCL Table file.");
+      auto [SecondWord, ThirdWord] = Rem1.split("|");
+      Entry.PropFile = SecondWord.str();
+      Entry.SymFile = ThirdWord.str();
+      Entries.push_back(Entry);
+    }
+    return Error::success();
+  }
+};
+
+// Run LLVM to SPIR-V translation.
+static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef InputTable,
+                                                     const ArgList &Args) {
+  Table LiveSYCLTable;
+  Expected<std::string> LLVMToSPIRVPath =
+      findProgram("llvm-spirv", {getMainExecutable("llvm-spirv")});
+  if (!LLVMToSPIRVPath)
+    return LLVMToSPIRVPath.takeError();
+
+  if (Error Err = LiveSYCLTable.populateSYCLTable(InputTable))
+    return std::move(Err);
+  auto InputFiles = LiveSYCLTable.getListOfIRFiles();
+
+  SmallVector<StringRef, 8> CmdArgs;
+  CmdArgs.push_back(*LLVMToSPIRVPath);
+  StringRef LLVMToSPIRVOptions;
+  if (Arg *A = Args.getLastArg(OPT_llvm_spirv_options_EQ))
+    LLVMToSPIRVOptions = A->getValue();
+  LLVMToSPIRVOptions.split(CmdArgs, " ", /* MaxSplit = */ -1,
+                           /* KeepEmpty = */ false);
+  CmdArgs.push_back("-o");
+
+  int I = 0;
+  for (const auto &File : InputFiles) {
```

Comment:
Nit: Maybe we could swap this so the loop is incrementing `I` and in the first line it indexes into `InputFiles` to assign I, instead of incrementing I outside of the loop